exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/common/HashTableTemplate.java

```diff
@@ -498,6 +497,13 @@ public void setup(HashTableConfig htConfig, BufferAllocator allocator, RecordBat
     currentIdxHolder = new IndexPointer();
   }
 
+  @Override
+  public void updateInitialCapacity(int initialCapacity) {
+    htConfig = htConfig.withInitialCapacity(initialCapacity);
+    allocationTracker = new HashTableAllocationTracker(htConfig, BATCH_SIZE);
+    enlargeEmptyHashTableIfNeeded(initialCapacity);
```

Comment:
@Ben-Zvi , @ilooner for typical foreign-key-primary-key joins, there is likely to be lots of duplicates of the foreign keys.  For example, in TPC-H, the o_custkey column in Orders table may have as many as 32 duplicates per value(see below).   If Orders is in the build side of a customer-orders hash join, the initial capacity of the hash table would be quite overestimated I think.  

+------------+-------------+
| o_custkey  | cust_count  |
+------------+-------------+
| 898        | 32          |
| 712        | 32          |
| 79         | 32          |
| 1282       | 32          |
| 643        | 32          |
| 4          | 31          |
| 1213       | 30          |
| 334        | 30          |
| 1078       | 30          |
| 73         | 30          |
+------------+-------------+

Reply:
Maybe in the future we could use stats to lower the "initial capacity" .  Now we could create a new special session option to reduce (e.g., divide by that option) this figure, but this has to be used manually, and would yield a small improvement, so better not.

Reply:
Let's create a JIRA to handle this (either through session option or some other way until proper stats are available).  Since there will be performance analysis needed to check the memory footprint, it can be done as a follow-up later.